spec/design/m-skel/AARCH64/MachineTypes.thy

```diff
@@ -52,16 +52,39 @@ text \<open>
 \<close>
 typedecl machine_state_rest
 
+type_synonym virq = machine_word
+
+text \<open> There are 64 general FPU registers saved. \<close>
+type_synonym fpu_regs = 64
+
+text \<open>
+  We use Haskell naming convention here, as we translate the Haskell FPUState directly
+  to this one for use in the abstract and executable specs.\<close>
+datatype fpu_state = FPUState (fpuRegs : "fpu_regs \<Rightarrow> 64 word")
+                              (fpuSr : "32 word")
+                              (fpuCr : "32 word")
+
 end
 
 qualify AARCH64 (in Arch)
 
-record
-  machine_state =
+record gic_vcpu_interface =
+  vgic_hcr  :: word32
+  vgic_vmcr :: word32
+  vgic_apr  :: word32
+  vgic_lr   :: "nat \<Rightarrow> AARCH64.virq"
+
+record vcpu_state =
+  vcpu_vgic :: "gic_vcpu_interface"
+  vcpu_regs :: "vcpureg \<Rightarrow> machine_word"
+
+record machine_state =
   irq_masks :: "AARCH64.irq \<Rightarrow> bool"
   irq_state :: nat
   underlying_memory :: "machine_word \<Rightarrow> word8"
   device_state :: "machine_word \<Rightarrow> word8 option"
+  vcpu_state :: vcpu_state
+  fpu_state :: AARCH64.fpu_state
```

Comment:
Now that https://github.com/seL4/l4v/pull/888 has been merged, does this still need to be qualified?

Reply:
Still necessary, this builds off that commit. Note that `fpu_state` (the type) has moved. It's similar to how `AARCH64.irq` is qualified